examples/consistency_models/train_consistency_distillation.py

```diff
@@ -0,0 +1,701 @@
+        disable=not accelerator.is_local_main_process,
+    )
+
+
+
+    # Train!
+    for epoch in range(first_epoch, args.num_epochs):
+        model.train()
+        for step, batch in enumerate(train_dataloader):
+            clean_images = batch["input"]
+            labels = batch["labels"]
+            # Sample noise that we'll add to the images
+            noise = torch.randn(clean_images.shape).to(clean_images.device)
+            # Sample a random timestep for each image, TODO - allow different timesteps in a batch
+            index = torch.randint(
+                0, noise_scheduler.config.num_train_timesteps-1,  (1,), device=clean_images.device
+            ).long()
+            timestep = timesteps[index]
+            timestep_prev = timestep + 1
+            # TODO, we should have an add noise in the scheduler maybe?
+            noised_image = clean_images + noise*append_dims(timestep, clean_images.ndim)
+            scaled_timesteps = noise_scheduler.scale_timestep(timestep)
+            scaled_timesteps_prev = noise_scheduler.scale_timestep(timestep_prev) 
+            target_model_ema.copy_to(target_model.parameters())
+
+            with accelerator.accumulate(model):
+                # Predict the noise residual
+                model_output = model(noised_image, scaled_timesteps, class_labels=labels).sample
+                distiller = noise_scheduler.step(
+                    model_output, timestep, noised_image, use_noise=False
+                ).prev_sample
+
+                # Heun Solver to get previous timestep image using teacher model
+                # TODO - make this cleaner
+                samples = noised_image
+                x = samples
+                model_output = teacher_model(x, scaled_timesteps, class_labels=labels).sample
+                teacher_denoiser = noise_scheduler.step(
+                    model_output, timestep, x, use_noise=False
+                ).prev_sample
+                d = (x - teacher_denoiser) / append_dims(scaled_timesteps, x.ndim)
+                samples = x + d * append_dims(scaled_timesteps_prev - scaled_timesteps, x.ndim)
+                model_output = teacher_model(samples, scaled_timesteps_prev, class_labels=labels).sample
+                teacher_denoiser = noise_scheduler.step(
+                    model_output, timestep_prev, samples, use_noise=False
+                ).prev_sample
+
+                next_d = (samples - teacher_denoiser) / append_dims(scaled_timesteps_prev, x.ndim)
+                denoised_image = x + (d + next_d) * append_dims((scaled_timesteps_prev - scaled_timesteps) /2, x.ndim)
+
+                # get output from target model
+                model_output = target_model(denoised_image, scaled_timesteps_prev, class_labels=labels).sample
+                distiller_target = noise_scheduler.step(
+                    model_output, timestep_prev, denoised_image, use_noise=False
+                ).prev_sample
```

Comment:
I think the gradients should be turned off here, analogously to how the gradients are turned off in [`target_denoise_fn`](https://github.com/openai/consistency_models/blob/main/cm/karras_diffusion.py#L127-L131). My understanding is that only the ("online") `model`''s parameters $`\theta`$ are updated using a backward pass, while the `target_model`'s parameters $`\theta^-`$ are only updated through the EMA update.